src/OmniSharp.Roslyn.CSharp/Services/Navigation/GotoDefinitionServiceV2.cs

```diff
@@ -46,15 +46,31 @@ public async Task<GotoDefinitionResponse> Handle(GotoDefinitionRequest request)
 
             if (symbol.Locations[0].IsInSource)
             {
-                return new GotoDefinitionResponse()
-                {
-                    Definitions = symbol.Locations
-                        .Select(location => new Definition
+                var definitions = symbol.Locations
+                    .Select(location =>
+                    {
+                        var sourceGeneratedFileInfo = GoToDefinitionHelpers.GetSourceGeneratedFileInfo(_workspace, location);
+                        var metadataSource = sourceGeneratedFileInfo == null && IsDecompiledSource(location.SourceTree) ?
```

Comment:
Since the `IsDecompiledSource` check looks pretty cheap and we expect `GetSourceGeneratedFileInfo` to return null for metadata SourceTrees, maybe we can avoid performing it by checking first.

Reply:
Something like the following?

```C#
                var definitions = symbol.Locations
                    .Select(location =>
                    {
                        MetadataSource? metadataSource = null;
                        SourceGeneratedFileInfo? sourceGeneratedFileInfo = null;

                        if (IsMetaDataSource(location.SourceTree))
                        {
                            metadataSource = new MetadataSource()
                            {
                                AssemblyName = symbol.ContainingAssembly.Name,
                                ProjectName = document.Project.Name,
                                TypeName = symbol.GetSymbolName()
                            };
                        }
                        else
                        {
                            sourceGeneratedFileInfo = GoToDefinitionHelpers.GetSourceGeneratedFileInfo(_workspace, location);
                        }
```

Reply:
That looks good to me.

Reply:
Changed and pushed. Cheers,